[jb] Use the default Java SDK version when there's no explicit configuration for it yet

felladrin opened the pull request:
## Description
<!-- Describe your changes in detail -->
Currently, when we open a Java Workspace, it selects Java 17 as SDK by default:
<img width="808" alt="image" src="https://user-images.githubusercontent.com/418083/185409355-6777c6c8-6640-444f-81e0-9864ecff455a.png">

This PR changes it to use the JDK preferred by the user (the ones defined in JAVA_HOME or PATH environment variables), in case it's not defined in a configuration file yet (e.g. `.idea/misc.xml`):
<img width="847" alt="image" src="https://user-images.githubusercontent.com/418083/185913994-3396fc20-ad01-4926-a441-ccc21f7baf87.png">

## How does it work

`jdkTable.preconfigure()` will set the `ProjectRootManager.getInstance(project).projectSdk` to a specific version if there’s an `.idea/misc.xml` file defining the JDK for the project.
And if there’s no configuration file, it will add a JDK to the list of available SDKs but won't set up the project SDK. That's where we set them up with the default JDK version found by `JavaHomeFinder.`

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes #12030

## How to test
<!-- Provide steps to test this PR -->
1. Open the preview environment generated for this branch
2. Choose the _Stable_ version of IntelliJ IDEA as your preferred editor
3. Test a repository that contains the JDK specified in `.idea/misc.xml`:
    a. Start a workspace using this repository: https://github.com/gitpod-io/spring-petclinic
    b. Verify that the workspace starts successfully
    c. Verify that the IDE opens successfully
    d. In JetBrains Client, open the menu "File" >> "Project Structure" and check if it's using the same JDK version as the one defined in `.idea/misc.xml`.
    <img width="803" alt="image" src="https://user-images.githubusercontent.com/418083/185628420-90296a5d-2817-4520-8d22-149827c44bee.png">
    e. Stop the workspace and restart it. Now confirm there are no duplicated SDKs listed in `File >> Project Structure >> Platform Settings >> SDKs`.
4. Test a repository that doesn't contain the JDK specified in `.idea/misc.xml` or doesn't contain that file at all:
    a. Start a workspace using this repository: https://github.com/gitpod-io/template-java-spring-boot
    b. Verify that the workspace starts successfully
    c. Verify that the IDE opens successfully
    d. In JetBrains Client, open the menu "File" >> "Project Structure" and check if it's using the same JDK version as the one you see when you run `echo $JAVA_HOME` on the terminal. (Notice there are also other auto-detected JDKs listed, allowing users to change it if they want.)
    <img width="847" alt="image" src="https://user-images.githubusercontent.com/418083/185913994-3396fc20-ad01-4926-a441-ccc21f7baf87.png">
    e. Stop the workspace and restart it. Now confirm there are no duplicated SDKs listed in `File >> Project Structure >> Platform Settings >> SDKs`.
    <img width="266" alt="image" src="https://user-images.githubusercontent.com/418083/185914963-4f7fb55f-c601-43ad-bfb6-74fa6b2f9a18.png">

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user-facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
JetBrains IDEs will use the default Java SDK version when there's no explicit configuration for it yet.
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, there is nothing to do here.
-->
None.

## Werft options:
<!--
Optional annotations to add to the werft job.

* with-preview - whether to create a preview environment for this PR
-->
- [x] /werft with-preview

felladrin commented:
Strange, it's behaving differently from when running via `launch-dev-server.sh`.
Threw this error:
```
java.lang.Throwable
	at com.intellij.util.indexing.RootChangesLogger.info(RootChangesLogger.java:27)
	at com.intellij.util.indexing.EntityIndexingServiceImpl.logRootChanges(EntityIndexingServiceImpl.java:108)
	at com.intellij.util.indexing.EntityIndexingServiceImpl.indexChanges(EntityIndexingServiceImpl.java:87)
	at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.synchronizeRoots(ProjectRootManagerComponent.java:299)
	at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.fireRootsChangedEvent(ProjectRootManagerComponent.java:212)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.fireRootsChanged(ProjectRootManagerImpl.java:477)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$1.fireRootsChanged(ProjectRootManagerImpl.java:137)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$1.fireRootsChanged(ProjectRootManagerImpl.java:134)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$BatchSession.rootsChanged(ProjectRootManagerImpl.java:108)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.makeRootsChange(ProjectRootManagerImpl.java:449)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge.fireRootsChanged(ProjectRootManagerBridge.kt:160)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge.access$fireRootsChanged(ProjectRootManagerBridge.kt:34)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge$JdkChangeListener.jdkAdded(ProjectRootManagerBridge.kt:254)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:649)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:629)
	at com.intellij.util.messages.impl.MessageBusImplKt.deliverMessage(MessageBusImpl.kt:399)
	at com.intellij.util.messages.impl.MessageBusImplKt.pumpWaiting(MessageBusImpl.kt:378)
	at com.intellij.util.messages.impl.MessageBusImplKt.access$pumpWaiting(MessageBusImpl.kt:1)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:437)
	at jdk.proxy1/jdk.proxy1.$Proxy110.jdkAdded(Unknown Source)
	at com.intellij.openapi.projectRoots.impl.ProjectJdkTableImpl.addJdk(ProjectJdkTableImpl.java:247)
	at com.intellij.openapi.projectRoots.impl.JavaAwareProjectJdkTableImpl.lambda$preconfigure$0(JavaAwareProjectJdkTableImpl.java:45)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1023)
	at com.intellij.openapi.projectRoots.impl.JavaAwareProjectJdkTableImpl.preconfigure(JavaAwareProjectJdkTableImpl.java:45)
	at io.gitpod.jetbrains.remote.GitpodProjectManager.configureSdks$lambda-1$lambda-0(GitpodProjectManager.kt:52)
```

Before printing `WARN - #k.c.CoroutineScope - gitpod: 'spring-petclinic' project: SDK detected: 17 (/home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu)`

<img width="1020" alt="image" src="https://user-images.githubusercontent.com/418083/185075773-2c96d554-c84f-46a7-abeb-b5bea98230e9.png">

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/418083/185075753-7d78dab3-fe89-42f4-a907-723824379b50.png">

Although `JAVA_HOME` is correctly set:
```
$ echo $JAVA_HOME
/home/gitpod/.sdkman/candidates/java/current

$ java --version
openjdk 11.0.16 2022-07-19 LTS
```

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -40,8 +42,16 @@ class GitpodProjectManager(
             application.runWriteAction {
                 try {
-                    ProjectJdkTable.getInstance().preconfigure()
-                    pendingSdk.complete(ProjectJdkTable.getInstance().allJdks.firstOrNull())
```
Should not it allredy return all jdks we need to loop and find one whcih is from java home?

There is also `JavaHomeFinder` can it be helpful? i.e. instead of trying to parse JAVA_HOME themself?  It seems to have different strategies: https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L41

felladrin replied:
> Should not it allredy return all jdks we need to loop and find one whcih is from java home?

Unfortunately, only Java 17 is in the list at the first start of the workspace (the other Detected SDKs only appear there after the project is already loaded) (so only at the restart of the workspace would the list contain both Java 17 and 11). I'm looking for a way to make this detection of sdks available to happen before we loop through `jdkTable.allJdks`.

> There is also JavaHomeFinder can it be helpful?

I tried it too, but it only useful info it returns is the homePath (as string) of each detect SDK.
There's no info about which one is marked as default as per `JAVA_HOME`.
That's why I'm still querying for the environment variable.

akosyakov replied:
> I tried it too, but it only useful info it returns is the homePath (as string) of each detect SDK. There's no info about which one is marked as default as per JAVA_HOME. That's why I'm still querying for the environment variable.

Looking at the code first one should be if you disable [[1](https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L33-L35)] these flags

felladrin replied:
> Looking at the code first one should be if you disable [[1](https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L33-L35)] these flags

Tried it too. Disabled all the flags but it was still returning an array with the 3 available SDK paths from `/home/gitpod/.sdkman/candidates/java/`.

Note also that none of the methods from `JavaHomeFinderBasic` return an SDK object. They all return paths, so I couldn't find a use for this class.

akosyakov replied:
Could you post values? It should lookup from PATH and then from JAVA_HOME. I would expect that PATH has default in too.

felladrin replied:
Yes, these are the logs returned by `JavaHomeFinder.getFinder().findExistingJdks().forEach { thisLogger().warn(it) }`
```
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/current
```
Yes, these are the logs returned by `JavaHomeFinder.getFinder().checkDefaultInstallDir(false).checkUsedInstallDirs(false).checkConfiguredJdks(false).findExistingJdks().forEach { thisLogger().warn(it) }`
```
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/current
```

But having these paths doesn't help in this case because it doesn't indicate which one is the one from $JAVA_HOME, and it also doesn't return an SDK object; only strings are returned.

I need an SDK object for adding it to the table `ProjectJdkTable.getInstance().addJdk(sdk)`, because I can't call `ProjectRootManager.getInstance(project).projectSdk = sdk` if it's not added to ProjectJdkTable first.

akosyakov replied:
I think taking first one here it a right choice because they are computed from PATH env var, i.e. the same way how to it is computed in the shell.

akosyakov replied:
/home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu is in fact correct

felladrin replied:
Gotcha. I'll take from this list instead of querying from the environment var then.

felladrin commented:
I've tried a lot of things, but I'm still facing this error (which I confirmed is also happening with the code from main branch, indicating it's due to some logic change on the latest IDEs).
```
2022-08-17 16:25:23,456 [  21755]   INFO - #c.i.u.i.RootChangesLogger - New rootsChanged event for "spring-petclinic" project with partial rescanning with trace_hash = 818578700:
java.lang.Throwable
	at com.intellij.util.indexing.RootChangesLogger.info(RootChangesLogger.java:27)
	at com.intellij.util.indexing.EntityIndexingServiceImpl.logRootChanges(EntityIndexingServiceImpl.java:108)
	at com.intellij.util.indexing.EntityIndexingServiceImpl.indexChanges(EntityIndexingServiceImpl.java:87)
	at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.synchronizeRoots(ProjectRootManagerComponent.java:299)
	at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.fireRootsChangedEvent(ProjectRootManagerComponent.java:212)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.fireRootsChanged(ProjectRootManagerImpl.java:477)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$1.fireRootsChanged(ProjectRootManagerImpl.java:137)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$1.fireRootsChanged(ProjectRootManagerImpl.java:134)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$BatchSession.rootsChanged(ProjectRootManagerImpl.java:108)
	at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.makeRootsChange(ProjectRootManagerImpl.java:449)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge.fireRootsChanged(ProjectRootManagerBridge.kt:160)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge.access$fireRootsChanged(ProjectRootManagerBridge.kt:34)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.project.ProjectRootManagerBridge$JdkChangeListener.jdkAdded(ProjectRootManagerBridge.kt:254)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:649)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:629)
	at com.intellij.util.messages.impl.MessageBusImplKt.deliverMessage(MessageBusImpl.kt:399)
	at com.intellij.util.messages.impl.MessageBusImplKt.pumpWaiting(MessageBusImpl.kt:378)
	at com.intellij.util.messages.impl.MessageBusImplKt.access$pumpWaiting(MessageBusImpl.kt:1)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:437)
	at jdk.proxy1/jdk.proxy1.$Proxy109.jdkAdded(Unknown Source)
	at com.intellij.openapi.projectRoots.impl.ProjectJdkTableImpl.addJdk(ProjectJdkTableImpl.java:247)
	at com.intellij.openapi.projectRoots.impl.JavaAwareProjectJdkTableImpl.lambda$preconfigure$0(JavaAwareProjectJdkTableImpl.java:45)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1023)
	at com.intellij.openapi.projectRoots.impl.JavaAwareProjectJdkTableImpl.preconfigure(JavaAwareProjectJdkTableImpl.java:45)
	at io.gitpod.jetbrains.remote.GitpodProjectManager.configureSdks$lambda-3$lambda-2$lambda-0(GitpodProjectManager.kt:61)
```

akosyakov commented:
> I've tried a lot of things, but I'm still facing this error (which I confirmed is also happening with the code from main branch, indicating it's due to some logic change on the latest IDEs).

Could you ask JB folks please?

felladrin commented:
It looks like it's working now!
<img width="921" alt="image" src="https://user-images.githubusercontent.com/418083/185215842-8cf73047-c71e-4ff0-be90-a4459f3a8ed2.png">
<img width="812" alt="image" src="https://user-images.githubusercontent.com/418083/185216186-268a0852-5944-4b32-879d-b77a4a71f6bd.png">
<img width="975" alt="image" src="https://user-images.githubusercontent.com/418083/185215983-d104f88f-0118-4594-a50b-65e84b64703d.png">

felladrin commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -29,20 +33,37 @@ class GitpodProjectManager(
+                        jdkTable.addJdk(sdk)
+                        pendingSdk.complete(sdk)
+                    } else {
+                        application.invokeAndWait { jdkTable.preconfigure() }
```
I added `application.invokeAndWait` cause I thought it would solve [this issue](https://github.com/gitpod-io/gitpod/pull/12163#issuecomment-1218251482).
As I confirmed the issue was unrelated to it, I'll remove it in my next cleanup-commit.

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -29,20 +33,37 @@ class GitpodProjectManager(
         configureSdks()
     }
 
-    /**
```
Could you please keep it

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -29,20 +33,37 @@ class GitpodProjectManager(
+                        pendingSdk.complete(jdkTable.allJdks.firstOrNull())
+                    }
                 } catch (t: Throwable) {
+                    application.invokeAndWait { jdkTable.preconfigure() }
```
What is for?

felladrin replied:
It was a mistake. Removed during the cleanup.

felladrin commented:
/werft run with-clean-slate-deployment=true

akosyakov commented:
@felladrin Actually for spring-petclinic template I won't expect it to work, since we already committed `.idea` and configured to use 17 instead: https://github.com/gitpod-io/spring-petclinic/blob/c08fce83e4f401c9ec7d940ed19e30961bcb5d51/.idea/misc.xml#L11

it is valid, i.e. if a user reconfigured a project to use another JDK and committed then our auto detection should not do anything. Only in a case if the project model is missing before. I think you need to use another repo without `.idea` for testing.

But it also sounds like we have a bug that our auto detection overrides whatever a user already configured and committed? Should we have a look into it now or it should be another PR?

felladrin commented:
@akosyakov, I've updated the code to address the issues you spotted. It now respects the user config if it exists. Can you please review it?

I've also updated the "How to test" section instructing the reviewer to test two different repositories.

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -65,17 +79,23 @@ class GitpodProjectManager(
+                    thisLogger().warn("gitpod: '${project.name}' project: " +
+                            "SDK was auto preconfigured: ${projectRootManager.projectSdk}")
+                } else {
+                    thisLogger().warn("gitpod: '${project.name}' project: " +
```
let remove it

felladrin replied:
Removed!

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -65,17 +79,23 @@ class GitpodProjectManager(
-                    thisLogger().warn("gitpod: '${module.name}' module: SDK was auto preconfigured: $sdk")
+                    thisLogger().warn("gitpod: '${module.name}' module: SDK was auto preconfigured: ${m.sdk}")
+                } else {
+                    thisLogger().warn("gitpod: '${module.name}' module: SDK was already configured: ${m.sdk}")
```
let remove it, it is unnecessary noise, or change to debug level that it it not visible in default output

felladrin replied:
Removed!

akosyakov commented on components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt:
```diff
@@ -65,17 +79,15 @@ class GitpodProjectManager(
                 val projectRootManager = ProjectRootManager.getInstance(project)
                 if (projectRootManager.projectSdk == null) {
                     projectRootManager.projectSdk = sdk
-                    thisLogger().warn("gitpod: '${project.name}' project: SDK was auto preconfigured: $sdk")
```
@felladrin I meant only to drop `SDK was already configured`, the fact when we really configure is useful

felladrin replied:
They're back now!

felladrin commented:
/werft run with-clean-slate-deployment=true

akosyakov commented:
/werft run

felladrin commented:
I’ve updated the code, and now it avoids creating duplicated JDKs.
I'll test it on the Preview Environment before marking it as Ready for Review again.

For reference, running on the dev server is working like this:

For https://github.com/gitpod-io/spring-petclinic:
```
gitpod: 'spring-petclinic' project: SDK detected: 17 (/home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu)
```

For https://github.com/spring-projects/spring-petclinic:
```
gitpod: 'spring-petclinic' project: SDK detected: zulu-11 (/home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu)
gitpod: 'spring-petclinic' module: SDK was auto preconfigured: zulu-11 (/home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu)
gitpod: 'spring-petclinic' project: SDK was auto preconfigured: zulu-11 (/home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu)
```

felladrin commented:
/werft run with-clean-slate-deployment=true

felladrin commented:
I found out the cause of https://github.com/spring-projects/spring-petclinic/ not connecting and not using the expected JDK:
- Their .gitpod.yml is using an specific Docker Image that has only Java 17 inside: https://github.com/spring-projects/spring-petclinic/commit/5aef5c380c564d9c22afef3e173b1ec6af2c54bd
- This docker image has broken SSH functionality from Gitpod :exploding_head: I can’t connect to https://github.com/spring-projects/spring-petclinic/ even on https://gitpod.io/ (the same behavior you saw on Preview Environment).
And I noticed it was only [working during development](https://github.com/gitpod-io/gitpod/pull/12163#issuecomment-1222120597) (`launch-dev-server.sh`) because it dosn’t make use of .gitpod.yml config when we open a repo in dev-server.

I've replaced it with this other repo: https://github.com/gitpod-io/template-java-spring-boot

It's now ready for review again.

akosyakov commented:
> This docker image has broken SSH functionality from Gitpod. I can’t connect to https://github.com/spring-projects/spring-petclinic/ even on https://gitpod.io/. And I noticed it was only https://github.com/gitpod-io/gitpod/pull/12163#issuecomment-1222120597 (launch-dev-server.sh) because it dosn’t make use of .gitpod.yml config when we open a repo in dev-serv

@felladrin could you create an issue for it and add into IDE project please? 🙏

akosyakov commented:
@andreafalzetti Could you help to verify it? 🙏

andreafalzetti commented:
/werft run with-clean-slate-deployment

andreafalzetti reviewed (approved):
Works as advertised!
